{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c8109bbc-6998-4006-b00a-f2a93a64f127",
   "metadata": {},
   "source": [
    "# Preprocessing of individual datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad8028d-85b2-4db3-a4cb-3421da5167fc",
   "metadata": {},
   "source": [
    "Import packages... \n",
    "- [scanpy](https://scanpy.readthedocs.io/en/stable/index.html)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ce71d39-2b08-4d41-b7d7-0543e0d05066",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import pandas as pd\n",
    "\n",
    "import gc\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb89d430-f6b0-4c41-a35c-ef0a792eb343",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORKDIR=\"/workdir/dwm269/scMuscle2\"\n",
    "os.chdir(WORKDIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "235b0f78-5f73-42cd-a50f-66cededed6ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scripts.py.scantils import *\n",
    "# from scripts.py.scanplots import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f70ce8b0-1285-44f6-ad7d-ad7ae157cab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc.enable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9add09aa-98af-49d2-a954-5c5377d8b53e",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATADIR=\"/workdir/dwm269/scMuscle2/data/align_out\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93abab66-d2c8-4573-94bd-18f2183b30a4",
   "metadata": {},
   "source": [
    "Scanpy settings info [[link](https://scanpy.readthedocs.io/en/stable/generated/scanpy._settings.ScanpyConfig.html)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cdf589d8-c809-448e-9d0c-7cde0d4df148",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.cachedir = './data/cache/'\n",
    "sc.settings.n_jobs=16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aba526dd-37fd-4dc9-bcff-6d302d6e0715",
   "metadata": {},
   "source": [
    "Load metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3bb981bb-0dc6-428f-a859-9d1286be39c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = pd.read_csv(\"scMuscle2_metadata_v1-0.csv\")\n",
    "meta = meta.loc[meta[\"include\"],] # remove unwanted metadata\n",
    "meta = meta[[x in [\"fastq\", \"bam\"] for x in meta[\"file.format\"]]] # remove samples w/ download issues\n",
    "\n",
    "meta = meta[[x in [\"muscle\", \"tendon\"] for x in meta[\"tissue\"]]] # subset by tissue \n",
    "meta = meta[[x in [\"Homo sapiens\"] for x in meta[\"species\"]]] # subset by species , \"Mus musculus\"\n",
    "meta = meta.iloc[[8,12,23,34,45],] # subset by row index\n",
    "\n",
    "meta = meta[[x != \"\" for x in meta[\"GSM.accession\"]]] # \n",
    "meta = meta[[x != \"\" for x in meta[\"sample\"]]] # \n",
    "\n",
    "meta.index = list(range(0, len(meta))) # reset row indices\n",
    "# meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a68dbb35-4bd1-4aff-808c-ce3d883a910a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "muscle    4\n",
       "tendon    1\n",
       "Name: tissue, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.tissue.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "aacf4679-94c8-40a3-8710-24d45a120424",
   "metadata": {},
   "source": [
    "Read in count data & initialize anndata objects. Also add metadata to each object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d7bf66d5-996e-4cf0-b477-a5f720c269a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try:\n",
    "#     scm_list.index\n",
    "# except:\n",
    "scm_list = pd.Series(\n",
    "    index = meta[\"GSM.accession\"],\n",
    "    data = [\"\"]*meta.shape[0]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e19412a5-75c8-4888-ab2e-1ef28905ef9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample: GSM5848681\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/utils.py:111: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-4', 'SNORD116-5']\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Loaded 5499 cells and 62703 genes...\n",
      "Sample: GSM4272893\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/utils.py:111: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-4', 'SNORD116-5']\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Loaded 2393 cells and 62703 genes...\n",
      "Sample: GSM5098738\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/utils.py:111: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-4', 'SNORD116-5']\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Loaded 7737 cells and 62703 genes...\n",
      "Sample: GSM5098749\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/utils.py:111: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-4', 'SNORD116-5']\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Loaded 3802 cells and 62703 genes...\n",
      "Sample: GSM4743496\n",
      "     Loaded 3963 cells and 62703 genes...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/utils.py:111: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-4', 'SNORD116-5']\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "for i in range(0,meta.shape[0]):\n",
    "    if os.path.exists(DATADIR+\"/\"+meta[\"GSM.accession\"][i]+\"/STARsolo/Solo.out/GeneFull/filtered/matrix.mtx.gz\"):\n",
    "        print(\"Sample: \" + meta[\"GSM.accession\"][i]) \n",
    "#         try:\n",
    "        scm_list[i] = sc.read_10x_mtx(\n",
    "                path=DATADIR+\"/\"+meta[\"GSM.accession\"][i]+\"/STARsolo/Solo.out/GeneFull/filtered\", \n",
    "                var_names='gene_symbols',\n",
    "                make_unique=True,\n",
    "                cache=True\n",
    "            )\n",
    "          \n",
    "        scm_list[i].var_names_make_unique()  # this is unnecessary if using `var_names='gene_ids'` in `sc.read_10x_mtx`\n",
    "\n",
    "        for j in range(0,meta.shape[1]): #metadata features\n",
    "            scm_list[i].obs[meta.columns[j]] = meta.iloc[i,j]\n",
    "\n",
    "        scm_list[i].layers['counts'] = scm_list[i].X # save counts as a layer for future plotting\n",
    "\n",
    "        print(\"     Loaded \" + str(scm_list[i].shape[0]) + \" cells and \" + str(scm_list[i].shape[1]) + \" genes...\")\n",
    "    else:\n",
    "        print(\"Can't find counts for \" + str(meta[\"GSM.accession\"][i]) + \"...\")\n",
    "#         except:\n",
    "#             print(\"Exception with \" + meta[\"sample\"][i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77190686-bba7-4420-8699-6ca9a06e4456",
   "metadata": {},
   "source": [
    "Add ambient-RNA-scrubbed counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9639b1a-1640-4714-8c65-39b74b7e0b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61a79f35-0607-4383-ae18-287d0345c010",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1422"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5d9260f-0c1e-4e08-b8c3-72a468e68159",
   "metadata": {},
   "source": [
    "QC filter and preprocess individual datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "661665b7-17ef-4cc8-81d7-9a475101a77a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM4743488: 4788 cells and 61860 features...\n",
      "After filtering: 3004 cells and 61860 features...\n",
      " \n",
      "GSM4743489: 3153 cells and 61860 features...\n",
      "After filtering: 1814 cells and 61860 features...\n",
      " \n",
      "GSM4743490: 3050 cells and 61860 features...\n",
      "After filtering: 2232 cells and 61860 features...\n",
      " \n",
      "GSM4743491: 2505 cells and 61860 features...\n",
      "After filtering: 1870 cells and 61860 features...\n",
      " \n",
      "GSM4743492: 2174 cells and 61860 features...\n",
      "After filtering: 1936 cells and 61860 features...\n",
      " \n",
      "GSM4743493: 2643 cells and 61860 features...\n",
      "After filtering: 2485 cells and 61860 features...\n",
      " \n",
      "GSM4743494: 2706 cells and 61860 features...\n",
      "After filtering: 2578 cells and 61860 features...\n",
      " \n",
      "GSM4743495: 3207 cells and 61860 features...\n",
      "After filtering: 2532 cells and 61860 features...\n",
      " \n",
      "GSM5848681: 5500 cells and 61860 features...\n",
      "After filtering: 5274 cells and 61860 features...\n",
      " \n",
      "GSM5848680: 5875 cells and 61860 features...\n",
      "After filtering: 5234 cells and 61860 features...\n",
      " \n",
      "GSM5848679: 6432 cells and 61860 features...\n",
      "After filtering: 6002 cells and 61860 features...\n",
      " \n",
      "GSM5848678: 10059 cells and 61860 features...\n",
      "After filtering: 9882 cells and 61860 features...\n",
      " \n",
      "GSM4272893: 2395 cells and 61860 features...\n",
      "After filtering: 1689 cells and 61860 features...\n",
      " \n",
      "GSM4272894: 902 cells and 61860 features...\n",
      "After filtering: 726 cells and 61860 features...\n",
      " \n",
      "GSM4272895: 1676 cells and 61860 features...\n",
      "After filtering: 1195 cells and 61860 features...\n",
      " \n",
      "GSM4272896: 382 cells and 61860 features...\n",
      "After filtering: 253 cells and 61860 features...\n",
      " \n",
      "GSM4272897: 2142 cells and 61860 features...\n",
      "After filtering: 921 cells and 61860 features...\n",
      " \n",
      "GSM4272898: 6601 cells and 61860 features...\n",
      "After filtering: 5615 cells and 61860 features...\n",
      " \n",
      "GSM4272899: 2799 cells and 61860 features...\n",
      "After filtering: 2565 cells and 61860 features...\n",
      " \n",
      "GSM4272900: 5430 cells and 61860 features...\n",
      "After filtering: 4845 cells and 61860 features...\n",
      " \n",
      "GSM4272901: 13524 cells and 61860 features...\n",
      "After filtering: 9551 cells and 61860 features...\n",
      " \n",
      "GSM4272902: 9916 cells and 61860 features...\n",
      "After filtering: 7760 cells and 61860 features...\n",
      " \n",
      "GSM5098737: 6440 cells and 61860 features...\n",
      "After filtering: 3587 cells and 61860 features...\n",
      " \n",
      "GSM5098738: 7748 cells and 61860 features...\n",
      "After filtering: 5203 cells and 61860 features...\n",
      " \n",
      "GSM5098739: 4335 cells and 61860 features...\n",
      "After filtering: 3030 cells and 61860 features...\n",
      " \n",
      "GSM5098740: 14670 cells and 61860 features...\n",
      "After filtering: 1556 cells and 61860 features...\n",
      " \n",
      "GSM5098741: 6059 cells and 61860 features...\n",
      "After filtering: 3208 cells and 61860 features...\n",
      " \n",
      "GSM5098742: 5780 cells and 61860 features...\n",
      "After filtering: 2695 cells and 61860 features...\n",
      " \n",
      "GSM5098743: 4232 cells and 61860 features...\n",
      "After filtering: 1451 cells and 61860 features...\n",
      " \n",
      "GSM5098744: 4037 cells and 61860 features...\n",
      "After filtering: 1515 cells and 61860 features...\n",
      " \n",
      "GSM5098745: 9432 cells and 61860 features...\n",
      "After filtering: 5177 cells and 61860 features...\n",
      " \n",
      "GSM5098746: 4778 cells and 61860 features...\n",
      "After filtering: 3750 cells and 61860 features...\n",
      " \n",
      "GSM5098747: 4128 cells and 61860 features...\n",
      "After filtering: 2692 cells and 61860 features...\n",
      " \n",
      "GSM5098748: 5410 cells and 61860 features...\n",
      "After filtering: 4066 cells and 61860 features...\n",
      " \n",
      "GSM5098749: 3799 cells and 61860 features...\n",
      "After filtering: 3041 cells and 61860 features...\n",
      " \n",
      "GSM5098750: 3429 cells and 61860 features...\n",
      "After filtering: 2037 cells and 61860 features...\n",
      " \n",
      "GSM5098751: 13305 cells and 61860 features...\n",
      "After filtering: 9348 cells and 61860 features...\n",
      " \n",
      "GSM5098752: 12866 cells and 61860 features...\n",
      "After filtering: 8155 cells and 61860 features...\n",
      " \n",
      "GSM5098753: 17689 cells and 61860 features...\n",
      "After filtering: 11387 cells and 61860 features...\n",
      " \n",
      "GSM5672476: 10032 cells and 61860 features...\n",
      "After filtering: 7799 cells and 61860 features...\n",
      " \n",
      "GSM5672477: 8612 cells and 61860 features...\n",
      "After filtering: 6644 cells and 61860 features...\n",
      " \n",
      "GSM5672478: 8448 cells and 61860 features...\n",
      "After filtering: 6673 cells and 61860 features...\n",
      " \n",
      "GSM5672479: 6448 cells and 61860 features...\n",
      "After filtering: 5622 cells and 61860 features...\n",
      " \n",
      "GSM5672480: 8972 cells and 61860 features...\n",
      "After filtering: 8467 cells and 61860 features...\n",
      " \n",
      "GSM5390515: 8297 cells and 61860 features...\n",
      "After filtering: 8127 cells and 61860 features...\n",
      " \n",
      "GSM4743496: 3938 cells and 61860 features...\n",
      "After filtering: 2273 cells and 61860 features...\n",
      " \n",
      "GSM5390456: 4017 cells and 61860 features...\n",
      "After filtering: 3788 cells and 61860 features...\n",
      " \n",
      "GSM5390460: 7282 cells and 61860 features...\n",
      "After filtering: 4597 cells and 61860 features...\n",
      " \n",
      "GSM5390463: 8770 cells and 61860 features...\n",
      "After filtering: 7909 cells and 61860 features...\n",
      " \n",
      "GSM5390474: 7369 cells and 61860 features...\n",
      "After filtering: 6617 cells and 61860 features...\n",
      " \n",
      "GSM5848676: 11025 cells and 61860 features...\n",
      "After filtering: 10495 cells and 61860 features...\n",
      " \n",
      "GSM5848675: 20391 cells and 61860 features...\n",
      "After filtering: 16981 cells and 61860 features...\n",
      " \n",
      "GSM5848674: 4659 cells and 61860 features...\n",
      "After filtering: 3559 cells and 61860 features...\n",
      " \n",
      "GSM5848673: 11505 cells and 61860 features...\n",
      "After filtering: 10724 cells and 61860 features...\n",
      " \n",
      "GSM5390475: 8261 cells and 61860 features...\n",
      "After filtering: 8091 cells and 61860 features...\n",
      " \n",
      "GSM5390476: 9575 cells and 61860 features...\n",
      "After filtering: 7146 cells and 61860 features...\n",
      " \n",
      "GSM5390477: 4824 cells and 61860 features...\n",
      "After filtering: 4524 cells and 61860 features...\n",
      " \n",
      "GSM5390478: 5368 cells and 61860 features...\n",
      "After filtering: 5036 cells and 61860 features...\n",
      " \n",
      "GSM5390479: 5018 cells and 61860 features...\n",
      "After filtering: 4879 cells and 61860 features...\n",
      " \n",
      "GSM5390480: 5256 cells and 61860 features...\n",
      "After filtering: 4944 cells and 61860 features...\n",
      " \n",
      "GSM5390481: 10425 cells and 61860 features...\n",
      "After filtering: 7726 cells and 61860 features...\n",
      " \n",
      "GSM5390483: 6181 cells and 61860 features...\n",
      "After filtering: 6039 cells and 61860 features...\n",
      " \n",
      "GSM5390489: 5731 cells and 61860 features...\n",
      "After filtering: 5632 cells and 61860 features...\n",
      " \n",
      "GSM5390490: 5183 cells and 61860 features...\n",
      "After filtering: 5112 cells and 61860 features...\n",
      " \n",
      "GSM5848693: 6310 cells and 61860 features...\n",
      "After filtering: 5129 cells and 61860 features...\n",
      " \n",
      "GSM5848694: 6545 cells and 61860 features...\n",
      "After filtering: 3644 cells and 61860 features...\n",
      " \n",
      "GSM5848692: 15213 cells and 61860 features...\n",
      "After filtering: 15080 cells and 61860 features...\n",
      " \n",
      "GSM5848691: 10253 cells and 61860 features...\n",
      "After filtering: 10132 cells and 61860 features...\n",
      " \n",
      "GSM5848690: 7438 cells and 61860 features...\n",
      "After filtering: 6296 cells and 61860 features...\n",
      " \n",
      "GSM5848689: 6182 cells and 61860 features...\n",
      "After filtering: 5571 cells and 61860 features...\n",
      " \n",
      "GSM5848688: 16898 cells and 61860 features...\n",
      "After filtering: 14178 cells and 61860 features...\n",
      " \n",
      "GSM5848687: 6087 cells and 61860 features...\n",
      "After filtering: 5081 cells and 61860 features...\n",
      " \n",
      "GSM5848685: 13200 cells and 61860 features...\n",
      "After filtering: 8498 cells and 61860 features...\n",
      " \n",
      "GSM5848686: 18405 cells and 61860 features...\n",
      "After filtering: 15451 cells and 61860 features...\n",
      " \n",
      "GSM5848684: 20133 cells and 61860 features...\n",
      "After filtering: 19204 cells and 61860 features...\n",
      " \n",
      "GSM5848683: 8076 cells and 61860 features...\n",
      "After filtering: 7503 cells and 61860 features...\n",
      " \n",
      "GSM5848682: 5988 cells and 61860 features...\n",
      "After filtering: 4012 cells and 61860 features...\n",
      " \n",
      "GSM4979413: 2672 cells and 61860 features...\n",
      "After filtering: 2060 cells and 61860 features...\n",
      " \n",
      "GSM4979414: 2875 cells and 61860 features...\n",
      "After filtering: 2306 cells and 61860 features...\n",
      " \n",
      "GSM4979415: 3520 cells and 61860 features...\n",
      "After filtering: 2852 cells and 61860 features...\n",
      " \n",
      "GSM4979416: 2898 cells and 61860 features...\n",
      "After filtering: 2445 cells and 61860 features...\n",
      " \n",
      "GSM4979417: 3786 cells and 61860 features...\n",
      "After filtering: 3152 cells and 61860 features...\n",
      " \n",
      "GSM4979418: 2799 cells and 61860 features...\n",
      "After filtering: 2297 cells and 61860 features...\n",
      " \n",
      "Exception with nan\n",
      " \n",
      "Exception with nan\n",
      " \n",
      "Exception with nan\n",
      " \n",
      "Exception with nan\n",
      " \n",
      "Exception with nan\n",
      " \n"
     ]
    }
   ],
   "source": [
    "# gene/transcript, mito filters count filters\n",
    "for i in range(0,meta.shape[0]):\n",
    "    try:\n",
    "        print(meta[\"GSM.accession\"][i] + ': ' + str(scm_list[i].shape[0]) + \" cells and \" + str(scm_list[i].shape[1]) +' features...')  \n",
    "\n",
    "        # Hard filters for feature and UMI counts\n",
    "        sc.pp.filter_cells(\n",
    "            scm_list[i],\n",
    "            min_genes=500\n",
    "        )\n",
    "        sc.pp.filter_cells(\n",
    "            scm_list[i], \n",
    "            min_counts=1000\n",
    "        )\n",
    "\n",
    "        scm_list[i].var['mito'] = scm_list[i].var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            scm_list[i], \n",
    "            qc_vars=['mito'], \n",
    "            percent_top=None, \n",
    "            log1p=False, \n",
    "            inplace=True\n",
    "        )  \n",
    "\n",
    "        # QC filter(s)\n",
    "        scm_list[i] = scm_list[i][scm_list[i].obs.pct_counts_mito < 40, :] \n",
    "\n",
    "        print('After filtering: ' + str(scm_list[i].shape[0]) + \" cells and \" + str(scm_list[i].shape[1]) +' features...')  \n",
    "    \n",
    "    except:\n",
    "            print(\"Exception with \" + str(meta[\"GSM.accession\"][i]))\n",
    "        \n",
    "    print(\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e293663-931f-4c88-b496-4c86432340f1",
   "metadata": {},
   "source": [
    "Add metadata to each object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "428b8836-81e7-4884-8334-ec2003f74ab9",
   "metadata": {},
   "source": [
    "## Doublet removal via Scrublet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecef4e5-21b2-4d5d-8dec-0621f8e1ac11",
   "metadata": {},
   "source": [
    "Estimate doublet scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a473a94-49d1-43c1-bb10-4898c1fa1a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sc.settings.set_figure_params(\n",
    "#     fontsize=8\n",
    "# )\n",
    "\n",
    "# for i in range(0,meta.shape[0]):\n",
    "#     try:\n",
    "#         sc.external.pp.scrublet(\n",
    "#             scm_list[i]\n",
    "#         )\n",
    "#         sc.external.pl.scrublet_score_distribution(\n",
    "#             scm_list[i],\n",
    "#             figsize =[6,2.25]\n",
    "#         )\n",
    "            \n",
    "#     except:\n",
    "#             print(\"Exception with \" + meta[\"sample\"][i])\n",
    "        \n",
    "#     print(\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b9e5357-3831-49fc-847b-a4af38762895",
   "metadata": {},
   "source": [
    "Estimate doublet score cutoff values for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "085d0cac-6c9a-400d-ad21-add8b1f4ad72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cutoff_threshold = [\n",
    "#     0.63, 0.63, \n",
    "#     0.59, #D4_200um\n",
    "#     0.52, \n",
    "#     0.6, #D4_1000um\n",
    "#     0.24, \n",
    "#     0.2, 0.24,\n",
    "# #     0.58, # D20_600um\n",
    "#     0.63, # D21_200um\n",
    "#     0.18,\n",
    "#     0.18 #D21_1000um\n",
    "# ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f1e3b5-b5c3-4cb5-bde5-22fd60e951c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scm_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9640d32-0093-4ad5-8c6d-13e0a9fa62be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(\"Final cell & feature counts:\\n\")\n",
    "# for i in range(0,meta.shape[0]):    \n",
    "#     scm_list[i] = scm_list[i][scm_list[i].obs[\"doublet_score\"] < cutoff_threshold[i],]\n",
    "#     print(meta[\"sample\"][i] + ': ' + str(scm_list[i].shape[0]) + \" cells and \" + str(scm_list[i].shape[1]) +' features...')  \n",
    "#     print(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10c272ea-0599-4ad5-a296-54ab8c9c085a",
   "metadata": {},
   "source": [
    "## Merge into a single AnnData object and preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7fe7f58b-3a59-47f8-932b-c91f9a6aabb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/_core/anndata.py:1785: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  [AnnData(sparse.csr_matrix(a.shape), obs=a.obs) for a in all_adatas],\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23394, 62703)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwm269/miniconda3/envs/scm2/lib/python3.10/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "scm = scm_list[0].concatenate(\n",
    "    scm_list[1:],\n",
    "    index_unique=None\n",
    "#     join=\"inner\"\n",
    "#     batch_key=\"sample\",\n",
    "    # batch_categories=meta[\"GSM.accession\"]\n",
    ")\n",
    "scm.obs_names_make_unique()\n",
    "print(scm.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d72af68c-8b61-4fe1-a060-8311d168c3b9",
   "metadata": {},
   "source": [
    "QC plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "14474503-afb1-4e5c-bf01-86f7dbd742cd",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"Could not find keys '['pct_counts_mito']' in columns of `adata.obs` or in adata.var_names.\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 6\u001b[0m\n\u001b[1;32m      1\u001b[0m sc\u001b[39m.\u001b[39msettings\u001b[39m.\u001b[39mset_figure_params(\n\u001b[1;32m      2\u001b[0m     fontsize\u001b[39m=\u001b[39m\u001b[39m8\u001b[39m,\n\u001b[1;32m      3\u001b[0m     figsize\u001b[39m=\u001b[39m[\u001b[39m18\u001b[39m,\u001b[39m4\u001b[39m]\n\u001b[1;32m      4\u001b[0m )\n\u001b[0;32m----> 6\u001b[0m sc\u001b[39m.\u001b[39;49mpl\u001b[39m.\u001b[39;49mviolin(\n\u001b[1;32m      7\u001b[0m     scm,\n\u001b[1;32m      8\u001b[0m     size\u001b[39m=\u001b[39;49m\u001b[39m0.5\u001b[39;49m,\n\u001b[1;32m      9\u001b[0m     rotation\u001b[39m=\u001b[39;49m\u001b[39m90\u001b[39;49m,\n\u001b[1;32m     10\u001b[0m \u001b[39m#     log=True,\u001b[39;49;00m\n\u001b[1;32m     11\u001b[0m     keys\u001b[39m=\u001b[39;49m[\u001b[39m'\u001b[39;49m\u001b[39mpct_counts_mito\u001b[39;49m\u001b[39m'\u001b[39;49m],\u001b[39m# '',''],\u001b[39;49;00m\n\u001b[1;32m     12\u001b[0m     stripplot\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m     13\u001b[0m     groupby\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mGSM.accession\u001b[39;49m\u001b[39m'\u001b[39;49m\n\u001b[1;32m     14\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/scm2/lib/python3.10/site-packages/scanpy/plotting/_anndata.py:769\u001b[0m, in \u001b[0;36mviolin\u001b[0;34m(adata, keys, groupby, log, use_raw, stripplot, jitter, size, layer, scale, order, multi_panel, xlabel, ylabel, rotation, show, save, ax, **kwds)\u001b[0m\n\u001b[1;32m    763\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    764\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mExpected number of y-labels to be `\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mlen\u001b[39m(keys)\u001b[39m}\u001b[39;00m\u001b[39m`, \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    765\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mfound `\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mlen\u001b[39m(ylabel)\u001b[39m}\u001b[39;00m\u001b[39m`.\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    766\u001b[0m     )\n\u001b[1;32m    768\u001b[0m \u001b[39mif\u001b[39;00m groupby \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m--> 769\u001b[0m     obs_df \u001b[39m=\u001b[39m get\u001b[39m.\u001b[39;49mobs_df(adata, keys\u001b[39m=\u001b[39;49m[groupby] \u001b[39m+\u001b[39;49m keys, layer\u001b[39m=\u001b[39;49mlayer, use_raw\u001b[39m=\u001b[39;49muse_raw)\n\u001b[1;32m    770\u001b[0m     \u001b[39mif\u001b[39;00m kwds\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39mpalette\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mNone\u001b[39;00m) \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    771\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_categorical_dtype(adata\u001b[39m.\u001b[39mobs[groupby]):\n",
      "File \u001b[0;32m~/miniconda3/envs/scm2/lib/python3.10/site-packages/scanpy/get/get.py:272\u001b[0m, in \u001b[0;36mobs_df\u001b[0;34m(adata, keys, obsm_keys, layer, gene_symbols, use_raw)\u001b[0m\n\u001b[1;32m    269\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    270\u001b[0m     alias_index \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m--> 272\u001b[0m obs_cols, var_idx_keys, var_symbols \u001b[39m=\u001b[39m _check_indices(\n\u001b[1;32m    273\u001b[0m     adata\u001b[39m.\u001b[39;49mobs,\n\u001b[1;32m    274\u001b[0m     var\u001b[39m.\u001b[39;49mindex,\n\u001b[1;32m    275\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39mobs\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    276\u001b[0m     keys,\n\u001b[1;32m    277\u001b[0m     alias_index\u001b[39m=\u001b[39;49malias_index,\n\u001b[1;32m    278\u001b[0m     use_raw\u001b[39m=\u001b[39;49muse_raw,\n\u001b[1;32m    279\u001b[0m )\n\u001b[1;32m    281\u001b[0m \u001b[39m# Make df\u001b[39;00m\n\u001b[1;32m    282\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(index\u001b[39m=\u001b[39madata\u001b[39m.\u001b[39mobs_names)\n",
      "File \u001b[0;32m~/miniconda3/envs/scm2/lib/python3.10/site-packages/scanpy/get/get.py:167\u001b[0m, in \u001b[0;36m_check_indices\u001b[0;34m(dim_df, alt_index, dim, keys, alias_index, use_raw)\u001b[0m\n\u001b[1;32m    165\u001b[0m         not_found\u001b[39m.\u001b[39mappend(key)\n\u001b[1;32m    166\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(not_found) \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m--> 167\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(\n\u001b[1;32m    168\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCould not find keys \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mnot_found\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m in columns of `adata.\u001b[39m\u001b[39m{\u001b[39;00mdim\u001b[39m}\u001b[39;00m\u001b[39m` or in\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    169\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00malt_repr\u001b[39m}\u001b[39;00m\u001b[39m.\u001b[39m\u001b[39m{\u001b[39;00malt_search_repr\u001b[39m}\u001b[39;00m\u001b[39m.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    170\u001b[0m     )\n\u001b[1;32m    172\u001b[0m \u001b[39mreturn\u001b[39;00m col_keys, index_keys, index_aliases\n",
      "\u001b[0;31mKeyError\u001b[0m: \"Could not find keys '['pct_counts_mito']' in columns of `adata.obs` or in adata.var_names.\""
     ]
    }
   ],
   "source": [
    "sc.settings.set_figure_params(\n",
    "    fontsize=8,\n",
    "    figsize=[18,4]\n",
    ")\n",
    "\n",
    "sc.pl.violin(\n",
    "    scm,\n",
    "    size=0.5,\n",
    "    rotation=90,\n",
    "#     log=True,\n",
    "    keys=['pct_counts_mito'],# '',''],\n",
    "    stripplot=False,\n",
    "    groupby='GSM.accession'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5918c928-6b64-4eed-b480-1712ad150f3c",
   "metadata": {},
   "source": [
    "Preprocessing..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c30b15df-b6d6-4978-bf24-0fea92717e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(\n",
    "    scm, \n",
    "    target_sum=1e4\n",
    ")\n",
    "\n",
    "sc.pp.log1p(\n",
    "    scm\n",
    ")\n",
    "\n",
    "sc.pp.highly_variable_genes(\n",
    "    scm,\n",
    "    subset=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33b2dd68-3953-4914-b027-34e278232cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.scale(\n",
    "    scm\n",
    ")\n",
    "sc.tl.pca(\n",
    "    scm, \n",
    "    svd_solver='arpack',\n",
    "    n_comps=50\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa14d1cd-3e9a-4e2d-9bcd-43607cb13a9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "scm.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "391c0549-4906-414f-98ac-d4baad9861b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(\n",
    "    scm,\n",
    "    n_pcs = npcs(scm, reduction=\"X_pca\"),\n",
    "    n_neighbors = 50\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d712051-d989-4b06-b333-b30190a6e8b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(\n",
    "    scm\n",
    "#     group_by =\"neighbors\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c06f77-38cb-4f62-9d16-65eba2fb5a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(\n",
    "    figsize=(6,6),\n",
    "    dpi=200\n",
    ")\n",
    "sc.pl.umap(\n",
    "    scm,\n",
    "    legend_loc = None,\n",
    "    color =[\"GSM.accession\"] #GSM.accession\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8473dbcb-0ddc-430a-a976-36b8fcf29638",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffebc67a-738b-4625-8885-280d8e02c40e",
   "metadata": {},
   "source": [
    "Add cell cycle inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef316b8f-79fe-4778-b1a1-b64e76b1ab54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in cell cycle genes lists\n",
    "s_genes = list(pd.read_csv(\"resources/gene_lists/seurat_s_genes_2019.csv\")['x'])\n",
    "g2m_genes = list(pd.read_csv(\"resources/gene_lists/seurat_g2m_genes_2019.csv\")['x'])\n",
    "\n",
    "# score and plot\n",
    "sc.tl.score_genes_cell_cycle(\n",
    "    scm, \n",
    "    s_genes=s_genes, \n",
    "    g2m_genes=g2m_genes, \n",
    "    use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa70c8bb-d553-44f4-ac69-7ba4ac43d0c4",
   "metadata": {},
   "source": [
    "Save merged anndata object for easy loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a0ac37f-fd23-4055-9c7f-fc1552056f23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save h5ad object for subsequent python analyses\n",
    "scm.write(\"data/pyobjs/homo_sapiens_v1.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scm2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "vscode": {
   "interpreter": {
    "hash": "c7291f77ceb1da3e6678d8e174be84a5cc1e2f9dc2317d6c8e55873d2d0ec6dc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
